File: pkg/vm-low-orbit/tests/fixtures/go/storage_content.go
//go:build storage_content

package main

//lint:file-ignore U1000 compiled file

import (
	"fmt"
	"io"

	"github.com/taubyte/go-sdk/event"
	"github.com/taubyte/go-sdk/storage"
)

var (
	expectedStorageCid = "bafybeidqnk6czrgcaydbxw54tf2qpjmd5pcefpeoudytygrbukfoaawo4i"
)

//export storageContentTest
func storageContentTest(e event.Event) uint32 {
	h, err := e.HTTP()
	if err != nil {
		return 1
	}

	if err := _storageContentTest(); err != nil {
		h.Write([]byte(fmt.Sprintf("storageContentTest failed with %v", err)))
		return 1
	}
	h.Write([]byte(`{"ping": "pong"}`))

	return 0
}

func _storageContentTest() error {
	data := []byte("Hello World")
	data2 := []byte(" Hello World AGAIN")

	content, err := storage.Create()
	if err != nil {
		return fmt.Errorf("failed creating new content with %v", err)
	}

	_, err = content.Write(data)
	if err != nil {
		return fmt.Errorf("failed writing with %v", err)
	}

	// Should fail since its at the end
	readData, err := io.ReadAll(content)
	if err != nil {
		return fmt.Errorf("read empty failed with %v", err)
	}
	if len(readData) > 0 {
		return fmt.Errorf("expected readData to be empty got %s", string(readData))
	}

	_, err = content.Seek(0, 0)
	if err != nil {
		return fmt.Errorf("failed seek with %v", err)
	}

	readData, err = io.ReadAll(content)
	if err != nil {
		return fmt.Errorf("failed first reading file with %v", err)
	}
	if string(readData) != string(data) {
		return fmt.Errorf("not matching string in read %s != %s", string(readData), string(data))
	}

	_, err = content.Write(data2)
	if err != nil {
		return fmt.Errorf("failed writing data to file with %v", err)
	}

	_, err = content.Seek(0, 0)
	if err != nil {
		return fmt.Errorf("failed seek 2 with %v", err)
	}

	readData, err = io.ReadAll(content)
	if err != nil {
		return fmt.Errorf("failed reading file with %v", err)
	}

	expectedLength := len(data) + len(data2)
	if len(readData) != expectedLength {
		return fmt.Errorf("read is not the same length as written %d != %d", len(readData), expectedLength)
	}

	cid, err := content.Push()
	if err != nil {
		return fmt.Errorf("failed pushing file with %v", err)
	}

	if cid.String() != expectedStorageCid {
		return fmt.Errorf("CID's do not match after push %s != %s", cid.String(), expectedStorageCid)
	}

	getContent, err := storage.Open(cid)
	if err != nil {
		return fmt.Errorf("failed getting cid %s with %v", cid.String(), err)
	}

	_, err = getContent.Seek(0, 0)
	if err != nil {
		return fmt.Errorf("failed seeking getContent wtih %v", err)
	}

	readData, err = io.ReadAll(getContent)
	if err != nil {
		return fmt.Errorf("failed last reading file with %v", err)
	}

	if len(readData) != len(data)+len(data2) {
		return fmt.Errorf("read is not the same length as written %d != %d", len(readData), len(data)+len(data2))
	}

	getCid, err := getContent.Cid()
	if err != nil {
		return fmt.Errorf("failed getting cid with %v", err)
	}

	if getCid.String() != expectedStorageCid {
		return fmt.Errorf("CID's do not match after open %s != %s", getCid.String(), cid.String())
	}

	return nil
}

File: pkg/vm-low-orbit/tests/fixtures/rust/srcs/storage_content.rs
extern crate taubyte_sdk;
use std::io::{Read, Seek, SeekFrom, Write};

use taubyte_sdk::{
    event::Event,
    http,
    storage::{Content, Storage},
};

#[no_mangle]
pub fn storageContentTest(event: Event) {
    let http = event.http().unwrap();

    let err = do_test(http);
    if err.is_err() {
        let err = http
            .write(format!("storageContentTest failed with: {}", err.err().unwrap()).as_bytes());
        if err.is_err() {
            println!("Err writing: {}", err.err().unwrap())
        }
    }
}

fn do_test(h: http::Event) -> Result<(), Box<dyn std::error::Error>> {
    let data = "Hello World";
    let data2 = " Hello World AGAIN";
    let expected_cid = "bafybeidqnk6czrgcaydbxw54tf2qpjmd5pcefpeoudytygrbukfoaawo4i";

    let mut content = Content::new()?;
    content.write_all(data.as_bytes())?;

    // Should fail since its at the end
    let mut buf = String::new();
    content.read_to_string(&mut buf)?;
    assert_eq!(buf, "");

    content.seek(SeekFrom::Start(0))?;

    let mut buf = String::new();
    content.read_to_string(&mut buf)?;
    assert_eq!(buf, data);

    content.write_all(data2.as_bytes())?;

    content.seek(SeekFrom::Start(0))?;

    let mut buf = String::new();
    content.read_to_string(&mut buf)?;
    assert_eq!(buf, (data.to_string() + data2));

    let cid = content.push()?;
    assert_eq!(cid.to_string(), expected_cid);
    content.close()?;

    let mut get_content = Content::open(cid)?;
    get_content.seek(SeekFrom::Start(0))?;

    let mut buf = String::new();
    get_content.read_to_string(&mut buf)?;
    assert_eq!(buf, (data.to_string() + data2));

    h.write("Success".as_bytes())?;
    Ok(())
}
